### ISSUE TITLE
ClientHandshaker currently only looking at first supported signature algorithm

### ISSUE THREAD

**Issue by j-hiden** [eclipse/californium#209]
Hi all,

I've encountered an issue in the scandium `ClientHandshaker.receivedServerHelloDone(...)` method. Basically I'm communicating with a server that returns 2x supported signature algorithms, the first one being SHA256withRSA and the second SHAwithECDSA. Currently the code only looks at the first entry and as I have everything set up to use SHAwithECDSA I am getting:

`InvalidKeyException: No installed provider supports this key: sun.security.ec.ECPrivateKeyImpl`

There is a TODO just above the line of code that is causing the issue.
I've been able to apply a "fix" (botch) locally to test that everything works (certs/keys etc) by iterating over `certificateRequest.getSupportedSignatureAlgorithms()` and catching Exceptions until a supported signature is found. Obviously this isn't ideal but as I'm pretty new to the library I'm unsure how to check this properly.

Any ideas on how to put a proper fix in place?

Cheers

**Comment by sophokles73** [comment]
Well, you are putting a finger into one of the (many) open wounds we still have on Scandium ;-)
Scandium has been implemented with a focus on ECC (as you have already figured out). There currently is no code in Scandium that properly handles the different types of crypto algorithms (RSA, DSA etc) and their corresponding signature algorithms :-(

It would be helpful if you could point out the section of code that you suspect is causing problems and which handshake message we are talking about.

**Comment by boaks** [comment]
ClientHandshaker, line 521:

> 			// TODO make sure, that signature is supported
> 			SignatureAndHashAlgorithm signatureAndHashAlgorithm = certificateRequest.getSupportedSignatureAlgorithms().get(0);
> 			certificateVerify = new CertificateVerify(signatureAndHashAlgorithm, privateKey, handshakeMessages, session.getPeer());
> 

Just takes the first one ("get(0)"). May better check the list for a proper one.
(I'm currently try to understand, if it would be better to react on not matching sig algos just when receive the certificate request, but I'm still reading :-) ).

**Comment by j-hiden** [comment]
Yes @boaks has the correct line.

If it helps, the exact source of the error is on line 172 of the [CertificateVerify.setSigature(...)](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/CertificateVerify.java#L172) method.

My temporary 'fix' was to have this method rethrow the `InvalidKeyException` (currently it only logs it as severe and continues) which then allows me to catch it further up in the `ClientHandshaker.receivedServerHelloDone(...)` method:

			for (SignatureAndHashAlgorithm algorithm : certificateRequest.getSupportedSignatureAlgorithms()) {
				try {
					certificateVerify = new CertificateVerify(algorithm, privateKey, handshakeMessages, session.getPeer());
					break;
				} catch (Exception e) {
					//Signature is not supported
				}
			}

Obviously this isn't ideal though :)

**Comment by j-hiden** [comment]
Hi again.

So I've been having a bit more of a poke around and might have come up with something a little cleaner for this. Ideas below:

Replaced original code in `ClientHandshaker.receivedServerHelloDone(...)` method (line 552) with:
```java
// TODO make sure, that signature is supported
certificateVerify = createCertificateVerify(certificateRequest.getSupportedSignatureAlgorithms());
//Probably should be a null check on certificateVerify (& logging) here for cases where no supported algo is found
```

Additional methods for checking supported algorithms vs preferred cipher suites (currently in the `ClientHandshaker` class but checking could potentially be moved elsewhere - in `CipherSuite` itself maybe?)
```java
private CertificateVerify createCertificateVerify(List<SignatureAndHashAlgorithm> supportedAlgorithms) {
	for (SignatureAndHashAlgorithm supportedAlgorithm : certificateRequest.getSupportedSignatureAlgorithms()) {
		if (isSupported(supportedAlgorithm)) {
			return new CertificateVerify(supportedAlgorithm, privateKey, handshakeMessages, session.getPeer());
		}
	}
	return null; //Or could we use some kind of default instead?
}

private boolean isSupported(SignatureAndHashAlgorithm supportedAlgorithm) {
	for (CipherSuite preferredCipherSuite : preferredCipherSuites) {
		if (signatureAlgorithmMatches(supportedAlgorithm.getSignature(), preferredCipherSuite.getKeyExchange())
				&& hashAlgorithmMatches(supportedAlgorithm.getHash(), preferredCipherSuite.getMacName())) {
			return true;
		}
	}
	return false;
}

private boolean signatureAlgorithmMatches(SignatureAlgorithm signatureAlgorithm, KeyExchangeAlgorithm keyExchangeAlgorithm) {
	switch (signatureAlgorithm) {
	case ECDSA:
		return keyExchangeAlgorithm == KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN;
	case RSA:
		return keyExchangeAlgorithm == KeyExchangeAlgorithm.RSA || keyExchangeAlgorithm == KeyExchangeAlgorithm.DH_RSA || keyExchangeAlgorithm == KeyExchangeAlgorithm.DHE_RSA;
	case DSA:
		return keyExchangeAlgorithm == KeyExchangeAlgorithm.DH_DSS || keyExchangeAlgorithm == KeyExchangeAlgorithm.DHE_DSS;
	case ANONYMOUS:
		return keyExchangeAlgorithm == KeyExchangeAlgorithm.DH_ANON;
	default:
		//TODO need to handle KeyExchangeAlgorithm.PSK?
		return false;
	}
}

private boolean hashAlgorithmMatches(HashAlgorithm hashAlgorithm, String macName) {
	if (macName == null) {
		return false;
	}
	return macName.contains(hashAlgorithm.name());
}
```

Happy to set up a PR if this looks like the right direction to go but as mentioned I'm pretty new to the library & security stuff really isn't my area so please do flag any issues you see :)

**Comment by sophokles73** [comment]
Hi @j-hiden,

sorry for letting you down on this so far, I have a lot on my plate currently ...
My understanding of the problem you are running into is that Scandium currently does not consider the server's preference for creating the signature in the client's certificate_verify message.
For that, I think we do not need to check the client's preferred cipher suites but instead we need to make sure that the configured private key is compatible with one of the server's preferred signature and hash algorithms.
FMPOV we need to iterate over the algorithms indicated by the server in its certificate_request message until we find one whose `SignatureAndHashAlgorithm.getSignature()` is compatible with the client's configured private key. This can be done by matching against `PrivateKey.getAlgorithm()` with the special case of `EC` (as returned by `PrivateKey.getAlgorithm()`) matches `ECDSA`.

It is true though, that the client's private key also needs to be consistent with and appropriate for being used with the client's preferred cipher suites. However, that doesn't seem to be the problem here.

Does that make sense?

**Comment by j-hiden** [comment]
Hi @sophokles73 - no problem, I've been pretty busy myself! Your suggestion seems to be similar to what I had but instead checking the against the client PrivateKey (whereas I was checking vs preferredCipherSuites). I'll take a closer look when I get chance.

**Comment by sophokles73** [comment]
fixed on master in 986a7a881b42d924f9693228b57a776ee44c7514
fixed on 2.0.x in bafd03309f269af4a20aa9fa8245bb10d766b137
fixed on 1.0.x in 6bbe95c97f20f5e62e01831c69f239f36348aefe
